ipa-core/src/protocol/ipa_prf/mod.rs

```diff
@@ -84,18 +89,36 @@ where
     F: PrimeField + ExtendableField,
     Replicated<F>: Serializable,
 {
-    // TODO (richaj): Add shuffle either before the protocol starts or, after converting match keys to elliptical curve.
-    // We might want to do it earlier as that's a cleaner code
+    let shuffled = shuffle_inputs(ctx.narrow(&Step::Shuffle), input_rows).await?;
+    let mut prfd_inputs =
+        compute_prf_for_inputs(ctx.narrow(&Step::ConvertInputRowsToPrf), shuffled).await?;
+
+    prfd_inputs.sort_by(|a, b| a.prf_of_match_key.cmp(&b.prf_of_match_key));
+    // Chunk the incoming stream of records into stream of vectors of records with the same PRF
+    let mut input_stream = stream_iter(prfd_inputs);
+    let first_row = input_stream.next().await;
+    if first_row.is_none() {
+        return Ok(vec![]);
+    }
+    let first_row = first_row.unwrap();
+    let rows_chunked_by_user = chunk_rows_by_user(input_stream, first_row);
 
-    let prfd_inputs =
-        compute_prf_for_inputs(ctx.narrow(&Step::ConvertInputRowsToPrf), input_rows).await?;
+    let batches = rows_chunked_by_user.collect::<Vec<_>>().await;
 
-    let histogram = compute_histogram_of_users_with_row_count(&prfd_inputs);
+    let mut sorted = Vec::new();
+    let mut ctx_ts = ctx.clone();
+    for mut batch in batches {
```

Comment:
This part is really lame for now I am doing a sequential sort per pseudonym. Ideally we should be able to spawn multiple sorts per pseudonym maybe limit the number of sorts spawned to a LIMIT.